File: algo_dart/binary_tree/binary_node.go
package binarytree

type BinaryNode[E any] struct {
	Value      E
	LeftChild  *BinaryNode[E]
	RightChild *BinaryNode[E]
}

func (b *BinaryNode[E]) TraverseInOrder(action func(value E)) {
	if b.LeftChild != nil {
		b.LeftChild.TraverseInOrder(action)
	}
	action(b.Value)
	if b.RightChild != nil {
		b.RightChild.TraverseInOrder(action)
	}
}

func (b *BinaryNode[E]) TraversePreOrder(action func(value E)) {
	action(b.Value)
	if b.LeftChild != nil {
		b.LeftChild.TraversePreOrder(action)
	}
	if b.RightChild != nil {
		b.RightChild.TraversePreOrder(action)
	}
}

func (b *BinaryNode[E]) TraversePostOrder(action func(value E)) {
	if b.LeftChild != nil {
		b.LeftChild.TraversePostOrder(action)
	}
	if b.RightChild != nil {
		b.RightChild.TraversePostOrder(action)
	}
	action(b.Value)

}

type T any

func heightOfABinaryTree(node *BinaryNode[T]) int {
	if node == nil {
		return -1
	}
	return 1 + max(heightOfABinaryTree(node.LeftChild),
		heightOfABinaryTree(node.RightChild))

}

File: algo_dart/binary_tree/binary_node.dart
import 'dart:math';

class BinaryNode<T> {
  T? value;
  BinaryNode<T>? leftChild;
  BinaryNode<T>? rightChild;

  BinaryNode({this.value, this.leftChild, this.rightChild});

  @override
  String toString() {
    return _diagram(this);
  }

  String _diagram(
    BinaryNode<T>? node, [
    String top = '',
    String root = '',
    String bottom = '',
  ]) {
    if (node == null) {
      return '$root null\n';
    }
    if (node.leftChild == null && node.rightChild == null) {
      return '$root ${node.value}\n';
    }
    final a = _diagram(node.rightChild, '$top ', '$top ----', '$top |  ');
    final b = '$root${node.value}\n';
    final c = _diagram(
      node.leftChild,
      '$bottom|',
      '$bottom-----',
      '$bottom    ',
    );
    return '$a$b$c';
  }

  void traverseInOrder(void Function(T? value) action) {
    leftChild?.traverseInOrder(action);
    action(value);
    rightChild?.traverseInOrder(action);
  }

  void traversePreOrder(void Function(T? value) action) {
    action(value);
    leftChild?.traversePreOrder(action);
    rightChild?.traversePreOrder(action);
  }

  void traversePostOrder(void Function(T? value) action) {
    leftChild?.traversePostOrder(action);
    rightChild?.traversePostOrder(action);
    action(value);
  }
}

int heightOfBinaryTree<T>(BinaryNode<T>? node) {
  if (node == null) {
    return -1;
  }
  return 1 +
      max(
        heightOfBinaryTree(node.leftChild),
        heightOfBinaryTree(node.rightChild),
      );
}
